### ISSUE TITLE
fitz.open on empty file is unclear: "RuntimeError: cannot tell in file"

### ISSUE THREAD

**Issue by cbm755** [pymupdf/PyMuPDF#1601]
Make a zero-byte empty file:
```bash
touch i_am_empty.pdf
```

Try to open it in fitz:
```
>>> import fitz
>>> doc = fitz.open("i_am_empty.pdf")
mupdf: cannot recognize version marker
mupdf: cannot tell in file
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-477c6d4fb265> in <module>
----> 1 doc = fitz.open("i_am_empty.pdf")

~/.local/lib/python3.10/site-packages/fitz/fitz.py in __init__(self, filename, stream, filetype, rect, width, height, fontsize)
   3843         _fitz.Document_swiginit(
   3844             self,
-> 3845             _fitz.new_Document(
   3846                 filename, stream, filetype, rect, width, height, fontsize
   3847             ),

RuntimeError: cannot tell in file
```

This error doesn't really help.  Perhaps "Empty file" or "perhaps the input is not a PDF file" or similar?

**Comment by JorjMcKie** [comment]
All these are messages coming directly from the base library.
Admittedly, in both these cases they are extremely technical ... but true 😁.
If I were unfair, I could point you to MuPDF's bug tracker to request better understandable error messages there.
But you know me: I won't do that 😉.
The error messages displayed usually are accompanied by a handful of warnings which are suppressed from being displayed. They - together with the error message itself - land in `fitz.mupdf_warnings()`.
The full bunch of messages can always be retrieved via
```python
>>> print(fitz.TOOLS.mupdf_warnings())  # case: empty file
cannot recognize version marker
trying to repair broken xref
repairing PDF document
cannot tell in file
cannot tell in file
>>> 
```
```python
>>> print(fitz.TOOLS.mupdf_warnings())  # case: non-PDF
cannot recognize version marker
trying to repair broken xref
repairing PDF document
no objects found
>>>
```
You could also set a global parameter to request all messages shown, instead of just the error:
```python
>>> fitz.TOOLS.mupdf_display_warnings(True)
True
>>> doc = fitz.open("algebra.ipynb")  # case: non-PDF
mupdf: cannot recognize version marker
mupdf: trying to repair broken xref
mupdf: repairing PDF document
mupdf: no objects found
Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    doc = fitz.open("algebra.ipynb")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: no objects found
```
```python
>>> doc = fitz.open("pdf", b"")  # case: empty file
mupdf: cannot recognize version marker
mupdf: trying to repair broken xref
mupdf: repairing PDF document
mupdf: cannot tell in file
mupdf: cannot tell in file
Traceback (most recent call last):
  File "<pyshell#12>", line 1, in <module>
    doc = fitz.open("pdf", b"")  # case: empty file
  File "C:\Users\haral\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: cannot tell in file
```

**Comment by JorjMcKie** [comment]
I think there are two or three ways to address your complaint:
1. (you) set that global parameter to ensure seeing a larger error context
2. (you) or, similarly, wrap the open in a try/except and print the complete stack of messages in the except clause
3. (me) add some logic which recognizes common cases and appends better explanations where possible.

Let me check what I can do.

**Comment by JorjMcKie** [comment]
How about this:
```python
import fitz
doc = fitz.open("pdf", b"")  # case: empty file
mupdf: cannot recognize version marker
mupdf: cannot tell in file
mupdf: cannot tell in file
Traceback (most recent call last):
  File "<pyshell#1>", line 1, in <module>
    doc = fitz.open("pdf", b"")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python310\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: file seems to be empty

doc = fitz.open("pdf", b"abra kadabra")  # case: invalid PDF
mupdf: cannot recognize version marker
mupdf: no objects found
mupdf: no objects found
Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    doc = fitz.open("pdf", b"abra kadabra")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python310\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: broken PDF document
```

**Comment by JorjMcKie** [comment]
I think, I now have it in shape (no more cryptic intermediate MuPDF messages):
```python
import fitz
doc=fitz.open("pdf", b"")
Traceback (most recent call last):
  File "<pyshell#1>", line 1, in <module>
    doc=fitz.open("pdf", b"")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python310\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: cannot open empty file

doc=fitz.open("pdf", b"hello")
Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    doc=fitz.open("pdf", b"hello")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python310\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: broken PDF document
```
Analogous non-PDF cases should react like this (whether empty or invalid content):
```python
doc=fitz.open("xps", b"hello")
Traceback (most recent call last):
  File "<pyshell#3>", line 1, in <module>
    doc=fitz.open("xps", b"hello")
  File "C:\Users\haral\AppData\Local\Programs\Python\Python310\lib\site-packages\fitz\fitz.py", line 3862, in __init__
    _fitz.new_Document(
RuntimeError: empty or broken document
```

**Comment by cbm755** [comment]
I wonder if the "cannot tell in file" thing should be reported upstream?  Do you know what "tell" is?  (It doesn't make sense to me.)

**Comment by cbm755** [comment]
You're new exceptions looks great to me!  

But are RuntimeError's the right thing?  I think I understand that from the PoV of interfacing with external library (it was unexpected response form the library).  But from the point of view of the PyMuPDF user?  I'm not sure.  And I'm rubbish at designing Python exceptions, but my gut reaction here is something like ValueError?

**Comment by JorjMcKie** [comment]
> I wonder if the "cannot tell in file" thing should be reported upstream? Do you know what "tell" is? (It doesn't make sense to me.)

No, this is a correct diagnosis - although cryptic and maybe not the most relevant one in that situation. This standard C function for file objects returns the current relative read or write position.

**Comment by JorjMcKie** [comment]
As per the exception types:
There is some ambiguity here.
If you look at the definitions, neither of `RuntimeError`, nor `ValueError` quite is it.
I already switch to `FileNotFoundError` where applicable now within `fitz.open()`, which is some progress.
A wrong file structure and an empty document file should most adequately be classified as `IOError` in my mind. But this is now just an alias of `OSError` - which again does not sound right at all.
* `ValueError` is defined as "Raised when an operation or function receives an argument that has the right type but an inappropriate value, ..."  - I think not.
* `RuntimeError`: "Raised when an error is detected that doesn’t fall in any of the other categories. The associated value is a string indicating what precisely went wrong." This looks best all in all, doesn't it?

I should also add, that this special situation, when an open goes wrong, the number of possible causes is limited. And already now I must do a backwards interpretation of MuPDF-generated error text.
Throughout the rest of PyMuPDF, there is no way to do this kind of analysis. So any error raised by MuPDF (or my C code) elsewhere becomes a `RuntimeError`.
Nothing to feel particularly happy about. So far I have avoided creating my own exception class, but maybe I should invent one and call it `MuPDFError` or `FitzError`. But it just would be a different label and neither carry additional explanatory value - except that it would clarify the originator.

**Comment by JorjMcKie** [comment]
Forgot to mention, that an empty (PDF) document will raise `EOFError`, which really does fit. I detect this from the MuPDF error string containing "... cannot tell ...".
Unfortunately, MuPDF for other document types (XPS and such) does not try a `tell()` and always report the same error for empty files and ill-structured ones. So I generate a RuntimeError here.

**Comment by JorjMcKie** [comment]
Another unfortunate fact I realized when dealing with your request (and a related one a few days ago):
* MuPDF does not always perform plausibility / structure checks during open.
    - It does it for PDF. In this case we even can differentiate between empty file conditions and real structure issues.
    - ZIP archive based documents (XPS, EPUB, CBZ) are checked for valid ZIP structures. But an empty file cannot be told apart from structure issues.
    - Image file based documents are never checked during open. This only happens when accessing the content itself. So you can open a JPEG as a PNG no problem. Things are subsequently corrected even without any complaints - and thus without noticing it in PyMuPDF. Empty files are reported as "unknown image file format" - and again cannot be told apart from other errors.
    - SVG files also are treated much like other images. But here we have a real MuPDF issue: if the file is empty or corrupt, requesting the rectangle of the first (and only) page, brings down the interpreter. So I am using my own additional checks to detect this and reject the open.
    - PyMuPDF always asserts that opening from memory uses a stream with non-zero length. Done for all document types.
* MuPDF always accepts documents with HTML, XHTML and XML file extensions / mime types. No matter whether they are empty or have syntax issues.
    - XML mime types are accpted as being of mime type FictionBook2.
    - Converting such files to PDF does work for empty files or if there are a handful of arbitrary bytes.
    - Same is true for making pixmaps ... at least for page 0.

**Comment by JorjMcKie** [comment]
**_Update:_**

I finally decided to make my own Python exception: ``BadDocumentData`` for reporting open errors.
I also figured out how to **_always_** separate empty file issues from file structure problems.
The new version will document it like this:
![grafik](https://user-images.githubusercontent.com/8290722/154848384-ed8ea9e6-b859-4967-be1c-2ffc701ed925.png)

**Comment by JorjMcKie** [comment]
I made another update, see below for the changes. I think it will become part of the next version like that. Thank you for bringing up this issue!

![grafik](https://user-images.githubusercontent.com/8290722/155105830-f2192fea-9f74-4ad8-a654-ebba74f441fd.png)

I have created pre-version [wheels](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse). Maybe you want to try out something.

**Comment by cbm755** [comment]
This looks pretty good to me!   So to use this, I need something like:

```
from fitz import EmptyFileError, FileDataError

...

try:
    doc = fitz.open(foo):
except ValueError, FileDataError, EmptyFileError:   # also RunTimeError if I want PyMuPDF <= 1.19.5
    # actions for bad files
except FileNotFoundError:
    # actions for not found files
```

Maybe EmptyFileError should be a subclass of FileDataError?  (note I didn't read code yet!)

**Comment by JorjMcKie** [comment]
Yes, you are right. You can do this (note: `"empty.txt"` is an empty file):
```python
In [1]: import fitz
In [2]: try:
   ...:     doc = fitz.open("empty.txt")
   ...: except fitz.FileDataError:
   ...:     print("empty files are a kind of file data errors")
empty files are a kind of file data errors
In [3]:
```
`ValueError` **_only_**  occurs if a file type was explicitely given, which is not one of MuPDF's supported document types. In all other cases `FileDataError` is raised.

So `doc = fitz.open("some.file")` will either work if a valid PDF, or else fail with `FileDataError`.
Similarly, `doc = fitz.open(stream=memory_area)` will work if a PDF or fail with `FileDataError`.
Each of the two potentially with sub exception ``EmptyFileError``.

Only the following will ever fail with a `ValueError`:
* `doc = fitz.open("file", filetype="xxx")`
* `doc = fitz.open("xxx", memory_area)`
* `doc = fitz.open(stream=memory_area, filetype="xxx")`

**Comment by JorjMcKie** [comment]
Why not do this:
```python
try:
    doc = fitz.open(somefile)
except Exception as e:
    etext = repr(e).split("(")[0]
    if etext == "FileNotFoundError":
        # ...
    elif etext == "EmptyFileError":
        # ...
    elif ...
```

**Comment by cbm755** [comment]
Both look reasonable.  I personally like the former.

> ValueError only occurs if a file type was explicitely given

Will this be clear from the docs?  (I just re-read it and yes I guess it is)

> Each of the two potentially with sub exception EmptyFileError.

I don't think this hierarchy was clear from the docs...?   How about:

```
Returns:
    ...
    FileDataError: tried to open a file with a broken format.  Most commonly
         ``"cannot open broken document"`` but other error messages of this
         type are possible (generated by MuPDF are possible.)
    EmptyFileError: a specific subclass of `FileDataError`; tried to open an
         empty file or stream.
```

(I hope I'm being helpful here and not just bikeshedding)

**Comment by JorjMcKie** [comment]
> I hope I'm being helpful here and not just bikeshedding

You definitely are, rest assured. I am glad you brought this up, because it definitely is an open end - and as indicated I always felt that dealing with exceptions is not state of the art.

I am now re-visiting this and I am confident that I found a way to select an adequate exception class on a per error level.

The last version of documentation I posted here, is still not the final one. I will try to be as explicit / precise as possible.

I am a bit concerned though:
Don't I break existing coding with those changes? Everything catching only `RuntimeError` and `ValueError` will miss the new possible cases.
What is your position on this?

**Comment by cbm755** [comment]
This is a good point!  Options off the top of my head:

  * wait until next non-patch version bump (~6 months?)
  * make your new errors subclasses of RuntimeError?  This way, folks doing `try: ... except RuntimeError: ...` would still have their code work.
      - you could then remove that during your next non-patch version bump.

Note: I have no experience with this.  First one sounds safe but boring ;-)

**Comment by JorjMcKie** [comment]
Ok, thanks for your thoughts!
Good idea that subclassing `RuntimeError`.
I even think it may stay like that forever: everyone picky enough can immediately check for subclass values. And the messages are already how they should be ...

**Comment by JorjMcKie** [comment]
Easy-peasy change:
```python
In [3]: try:
   ...:     doc = fitz.open("empty.txt")
   ...: except RuntimeError:
   ...:     print("caught as RuntimeError")
caught as RuntimeError
In [4]: try:
   ...:     doc = fitz.open("empty.txt")
   ...: except fitz.FileDataError:
   ...:     print("caught as FileDataError")
caught as FileDataError
In [5]: try:
   ...:     doc = fitz.open("empty.txt")
   ...: except fitz.EmptyFileError:
   ...:     print("caught as EmptyFileError")
caught as EmptyFileError
```
Now for a file with structure issues:
```python
In [7]: try:
   ...:     doc = fitz.open("invalid.pdf")
   ...: except RuntimeError:
   ...:     print("caught as RuntimeError")
caught as RuntimeError
In [8]: try:
   ...:     doc = fitz.open("invalid.pdf")
   ...: except fitz.FileDataError:
   ...:     print("caught as FileDataError")
caught as FileDataError
In [9]: try:
   ...:     doc = fitz.open("invalid.pdf")
   ...: except fitz.EmptyFileError:
   ...:     print("NOT caught!")
---------------------------------------------------------------------------
FileDataError                             Traceback (most recent call last)
Input In [9], in <module>
      1 try:
----> 2     doc = fitz.open("invalid.pdf")
      3 except fitz.EmptyFileError:
      4     print("NOT caught!")

...
   3888 _fitz.Document_swiginit(
   3889     self,
-> 3890     _fitz.new_Document(
   3891         filename, stream, filetype, rect, width, height, fontsize
   3892     ),
...

FileDataError: cannot open broken document
In [10]:
```

**Comment by JorjMcKie** [comment]
Thank you. I will go with this then. The documentation would look like this:

![grafik](https://user-images.githubusercontent.com/8290722/155593024-6b4c97fc-eb30-479f-b238-fc8e13d16f81.png)

**Comment by cbm755** [comment]
Looks good!  Minor typo: `issued ++by++ module os`

**Comment by JorjMcKie** [comment]
"bei" is a German word pronounced the same.

**Comment by JorjMcKie** [comment]
Fixed with new version 1.19.6
